### ISSUE TITLE
API: Resources uploaded via that API do not display image preview

### ISSUE THREAD

**Issue by linzjax** [Cadasta/cadasta-platform#1322]
### Steps to reproduce the error
- Find an aws link to a resource already uploaded
- send a post request to `/api/v1/organization/<blah>/projects/<blarg>/spatial/<blarf-id>/resources/ file=<aws_link> name='blarg' original_file='image.jpg' AUTHORIZATION:'Token <api key>'`

### Actual behavior
- Resource gets added to the location, but there's no preview image
- When you download it, it is the correct resource

<img width="610" alt="screen shot 2017-03-20 at 4 36 31 pm" src="https://cloud.githubusercontent.com/assets/8128188/24120557/67e9bf4a-0d8b-11e7-9892-bab846fb987e.png">

### Expected behavior
- Preview image is created.

**Comment by khantaalaman** [comment]
I would like to take this up.

I need help with two things in the link: 

1.  Generate AWS link?
2.  API Key?

Thanks!

**Comment by oliverroick** [comment]
For developing and local testing purposes you won't need an AWS access key. Have a look at the way we create thumbnails when uploading a file via the UI, it should give you a sufficient example how to approach it.

**Comment by rishabhiitbhu** [comment]
@linzjax could you explain me what do you mean by aws link, is it the link where uploaded resources get stored (like `/media/s3/uploads/resources...`) ?
I have solved this issue but I have another question suppose If we add this [image-link](https://cdn4.iconfinder.com/data/icons/new-google-logo-2015/400/new-google-favicon-512.png) as `file` in post request then also this issue occures. Should I solve this case also before opening my PR.

**Comment by niharika1995** [comment]
The problem occurs because the `mime_type` of the resource isn't set to `image/jpeg` anytime during the request-response cycle. In the web UI, the `mime_type` is set using code in `file-upload.js`.

IMO, the best solution to this problem would be to set `mime_type` of the _Resource_ object within its `save` method and trim it out of `file_upload.js`.
Should I proceed with this approach?

**Comment by rishabhiitbhu** [comment]
@niharika1995 you are correct that the problem occurs because the `mime_type` of resource isn't set while and after post request to api and in the web UI the `mime_type` is set by `file-upload.js`.

I have solved this issue using a different approach. While creating through post request to api, the _resource_ object is created by `create` method in `ResourceSerializer` class of `resource/serializers.py` and setting the `mime_type` there only.

**Comment by artiladdha** [comment]
I would like to work on this.

**Comment by oliverroick** [comment]
@artiladdha There's already a PR for this. You can work on this issue for the sake of applying for an internship but it's not very likely your work will end up in the platform

**Comment by artiladdha** [comment]
Then please suggest me issue which is not yet solved.

**Comment by linzjax** [comment]
FYI if others are still looking at this: For uploading resources to AWS we're using a custom library called [django-buckets](https://github.com/Cadasta/django-buckets). The s3/uploads/resources/ is a from the FakeS3Storage used to mimic an S3 library in production, but the actual URLs will be different. Part of the problem is that with the `file_upload.js` you're given the full file to work with, so all you have to do to get the `mime_type` is call file.type. Since the API uses django-buckets, the "file" you upload has to already be uploaded into AWS, so all you get is the link.

**Comment by niharika1995** [comment]
@linzjax Is there a particular drawback on simply guessing the mimetype based on the file extension (if I am not wrong, that is what happens within `file_upload.js`? I do not see why does there need to exist a db-field `mime_type` when it can be ascertained at runtime.

**Comment by amplifi** [comment]
@niharika1995 Deriving the mimetype has to be done cautiously, as it's very easy for a user to upload a malicious file with a fake extension or to pass a header via API that declares a mimetype that doesn't actually match the file. Any mimetype received needs to be verified before acceptance.

**Comment by amplifi** [comment]
Resolving this issue will require adding a secure process by which we derive and/or verify the mimetype of the uploaded file, saving mimetype to the resources table in the db (and writing to S3 metadata if not already populated), quarantining files for which a user-specified mimetype doesn't match the derived mimetype and applying additional processing, then generating and displaying the appropriate preview image. Much of this is adjacent to #490.

**Comment by amplifi** [comment]
Will be addressed in #490

**Comment by dpalomino** [comment]
There is a workaround for this, so decreasing priority for now.
